micro-sam napari plugin

GenevieveBuckley opened the pull request:
This is an initial implementation adding napari plugin bindings to the micro-sam package. 
@constantinpape and I talked about this earlier, since it is a nice convenience feature.

This PR:
* Adds the files needed to be recognised as a napari plugin (napari.yaml, and the setup.cfg & pyproject.toml files, etc.)
* Provides all of the micro-sam example datasets via the napari` File > Sample Data` menu.
* ...next, we can provide the micro-sam widgets via the napari `Plugins` menu (following [this example](https://github.com/napari/cookiecutter-napari-plugin/blob/6e1d851ad6a0e2b3a584366c4fdef3f3dcd433da/%7B%7Bcookiecutter.plugin_name%7D%7D/src/%7B%7Bcookiecutter.module_name%7D%7D/_widget.py#L37-L46))

You should be able to test it out the new micro-sam plugin in napari like this:
```
git clone https://github.com/computational-cell-analytics/micro-sam.git
cd micro-sam
gh pr checkout 165
pip install -e .
npe2 validate micro-sam
# ✔ Manifest for 'SegmentAnything for Microscopy' valid!
napari
# From the napari File menu, go to "Sample Data" and open 'image-series' or one of the other micro-sam samples
```

GenevieveBuckley commented on MANIFEST.in:
```diff
@@ -0,0 +1,5 @@
+include LICENSE
```
The napari cookiecutter example plugin is set up with:
* pyproject.toml
* setup.cfg (no setup.py file, most of that info is moved into setup.cfg)
* MANIFEST.in
* and a napari.yaml file.

Both the pyproject.toml and MANIFEST.in files are exactly the same as the [cookiecutter plugin example](https://github.com/napari/cookiecutter-napari-plugin).
I replaced the current setup.py file with a setup.cfg, and transferred the information across.

GenevieveBuckley commented on micro_sam/__init__.py:
```diff
@@ -5,4 +5,22 @@
 """
 
-from .__version__ import __version__
+__version__ = "0.1.2.post1"
```
I don't know why, but if I kept the line `from .__version__ import __version__` here then it would not install with pip. It seemed to be trying to import the whole micro-sam package instead of just the version string we wanted.

If you know how to avoid this problem, please let me know.

<details><summary>Details</summary>
```
$ pip install -e .

Obtaining file:///Users/genevieb/Documents/GitHub/micro-sam
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing editable metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [59 lines of output]
      Traceback (most recent call last):
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/expand.py", line 81, in __getattr__
          return next(
      StopIteration

      The above exception was the direct cause of the following exception:

      Traceback (most recent call last):
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/expand.py", line 193, in read_attr
          return getattr(StaticModule(module_name, spec), attr_name)
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/expand.py", line 87, in __getattr__
          raise AttributeError(f"{self.name} has no attribute {attr}") from e
      AttributeError: micro_sam has no attribute __version__

      During handling of the above exception, another exception occurred:

      Traceback (most recent call last):
        File "/Users/genevieb/mambaforge/envs/napari/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/genevieb/mambaforge/envs/napari/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/Users/genevieb/mambaforge/envs/napari/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 181, in prepare_metadata_for_build_editable
          return hook(metadata_directory, config_settings)
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 473, in prepare_metadata_for_build_editable
          return self.prepare_metadata_for_build_wheel(
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 396, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 341, in run_setup
          exec(code, locals())
        File "<string>", line 1, in <module>
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/__init__.py", line 107, in setup
          return distutils.core.setup(**attrs)
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 159, in setup
          dist.parse_config_files()
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/dist.py", line 895, in parse_config_files
          setupcfg.parse_configuration(
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/setupcfg.py", line 189, in parse_configuration
          meta.parse()
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/setupcfg.py", line 500, in parse
          section_parser_method(section_options)
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/setupcfg.py", line 475, in parse_section
          self[name] = value
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/setupcfg.py", line 293, in __setitem__
          parsed = self.parsers.get(option_name, lambda x: x)(value)
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/setupcfg.py", line 606, in _parse_version
          return expand.version(self._parse_attr(value, self.package_dir, self.root_dir))
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/setupcfg.py", line 417, in _parse_attr
          return expand.read_attr(attr_desc, package_dir, root_dir)
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/expand.py", line 196, in read_attr
          module = _load_spec(spec, module_name)
        File "/private/var/folders/82/phlx_f6j5518qtvpm4x0_0fm0000gq/T/pip-build-env-hwd34tc7/overlay/lib/python3.10/site-packages/setuptools/config/expand.py", line 216, in _load_spec
          spec.loader.exec_module(module)  # type: ignore
        File "<frozen importlib._bootstrap_external>", line 883, in exec_module
        File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
        File "/Users/genevieb/Documents/GitHub/micro-sam/./micro_sam/__init__.py", line 11, in <module>
          from .sample_data import (
        File "/Users/genevieb/Documents/GitHub/micro-sam/./micro_sam/sample_data.py", line 9, in <module>
          import imageio.v3 as imageio
      ModuleNotFoundError: No module named 'imageio'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```
<p>



</p>
</details>

constantinpape replied:
It usually works for me with
```
pip install -e .
```
but I haven't tried with the new `setup.cfg` yet. Will check it out and see if I can reproduce the problem.

GenevieveBuckley commented on micro_sam/sample_data.py:
```diff
@@ -36,6 +39,23 @@ def fetch_image_series_example_data(save_directory: Union[str, os.PathLike]) ->
+    # [(data1, add_image_kwargs1), (data2, add_image_kwargs2)]
+    # Check the documentation for more information about the
+    # add_image_kwargs
+    # https://napari.org/stable/api/napari.Viewer.html#napari.Viewer.add_image
```
This `sample_data.py` file was created

We won't have to do anything similar for the widgets - just adding type hints should be enough there. See [this example](https://github.com/napari/cookiecutter-napari-plugin/blob/6e1d851ad6a0e2b3a584366c4fdef3f3dcd433da/%7B%7Bcookiecutter.plugin_name%7D%7D/src/%7B%7Bcookiecutter.module_name%7D%7D/_widget.py#L37-L46)

GenevieveBuckley commented on setup.cfg:
```diff
@@ -0,0 +1,86 @@
+
+[options]
+packages = find:
+install_requires =
```
Many of these packages (python-elf, vigra, etc.) appear to be available on conda-forge but not PyPI.
I've commented the install requirement out right now, because I am doing a lot of `pip install -e .` while I test things. Ideally we should fix this part up properly.

Currently there are no requirements for these packages to be installed before micro-sam, you just run into import errors if they are not there.

constantinpape replied:
Indeed, some of my "stack" is not on pip. The underlying reasons are vigra and nifty:
- https://github.com/constantinpape/vigra
- https://github.com/DerThorsten/nifty

which are fairly complex C++ packages and hence not so easy to provide via pip.

What do you think is the best strategy here?

GenevieveBuckley replied:
I think the best strategy is to:
1. Distribute micro-sam to users via conda-forge
2. For developers, we put a small section in the docs somewhere that says how to make a development environment, and that will include what dependencies you need to install.

constantinpape commented:
Hi @GenevieveBuckley ,
this is amazing!! I will have a closer look laterl

constantinpape commented on micro_sam/__init__.py:
```diff
@@ -5,4 +5,22 @@
+    sample_data_tracking,
+)
+
+__all__ = (
```
Is it necessary to use the `__all__` here for the napari plugin? This might shadow other things in `pdoc` (I haven't checked yet, so not sure if it's actually a problem.)

GenevieveBuckley replied:
No, I don't think it is necessary at all. We can get rid of it if you like!

GenevieveBuckley replied:
Update: it is fine, I can get rid of this :)

constantinpape commented on micro_sam/sample_data.py:
```diff
@@ -21,7 +23,8 @@ def fetch_image_series_example_data(save_directory: Union[str, os.PathLike]) ->
     fname = "image-series.zip"
-    unpack_filenames = [os.path.join("series", f"im{i}.tif") for i in range(3)]
+    # use first two files for image series (thrid file is not the same shape)
+    unpack_filenames = [os.path.join("series", f"im{i}.tif") for i in range(2)]
```
I included the image of a different shape on purpose, because the image series annotator supports that.
Is there a particular reason you removed it here?

GenevieveBuckley replied:
Oh, I see. Thank you for explaining!
How do you want these to be opened in napari, as three separate layers?
Or I could put them all in the same layer if I pad with zeros to make them the same shape.

napari does not support different shaped image frames in the same layer well. Each napari layer contains a `.data` numpy array, and numpy does not support joining different array shapes like that (the [awkward array](https://awkward-array.org/doc/main/) project does support this, but I don't think anyone has ever tried to combine it with napari - I suspect it would be difficult)

constantinpape replied:
The series annotation tool loads these images one at a time. The idea is to enable users to process multiple images from a folder more conveniently.
Check out https://github.com/computational-cell-analytics/micro-sam/blob/master/examples/image_series_annotator.py to run the demo for it.

constantinpape replied:
And if it's a problem for napari that the image shapes are different than let's skip the plugin for this example data. I am not really sure if it makes sense at all to provide the image series annotator as a plugin, so we can skip this data for now.

GenevieveBuckley replied:
Ok, I'll remove the image series from the sample datasets provided to napari.

constantinpape commented on pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+build-backend = "setuptools.build_meta"
+
+
+[tool.black]
```
What is the black configuration in the toml for exactly? (I know `black`, but I am not sure why to put it here; I am not so familiar with the toml way of setting things up.)

GenevieveBuckley replied:
Everything in the `pyproject.toml` is copied directly from the [cookiecutter napari plugin](https://github.com/napari/cookiecutter-napari-plugin) example. So this is here because whoever set that up liked those formatting tools. We can absolutely remove this if you want.

GenevieveBuckley replied:
I realized this doesn't actually answer your question of "why put the configuration settings in the pyproject.toml file?"
The answer to that is it is potentially useful if you want configuration settings specific for this particular project (but here it's all very generic so perhaps doesn't make as much sense).
And yes, you could put the black config settings in their own separate file - I guess people prefer to put everything in the pyrpoject.toml because it can get messy with a lot of scattered config files if you use a lot of different tools together.

[Here is what the black docs have to say about when this might be useful](https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#configuration-via-a-file):
> Black is able to read project-specific default values for its command line options from a pyproject.toml file. This is especially useful for specifying custom --include and --exclude/--force-exclude/--extend-exclude patterns for your project.

constantinpape commented on pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+target-version = ['py38', 'py39', 'py310']
+
+
+[tool.ruff]
```
Same question as for `black`. (And I don't know `ruff`, but seems to be a linter as well?!)

GenevieveBuckley replied:
Same as above, everything in the `pyproject.toml` is copied directly from the [cookiecutter napari plugin](https://github.com/napari/cookiecutter-napari-plugin) example. So this is here because whoever set that up liked those formatting tools. We can absolutely remove this if you want.

constantinpape replied:
Thanks for explaining! It's not a big deal either way, let's leave it in for now, and I will decide what to do when we finalize the PR ;).

GenevieveBuckley replied:
Ok, I'll leave the black and ruff configuration settings in place for now.

constantinpape reviewed (commented):
I had a look through and left some comments, gonna give it a try later.

GenevieveBuckley commented:
You access the sample datasets from the napari File menu (but plugin widgets are accessed from the plugins menu).

File > Open Sample > SegmentAnything for Microscopy > ...

<img width="1440" alt="Screenshot-micro-sam-sample-data" src="https://github.com/computational-cell-analytics/micro-sam/assets/30920819/8e4b30b7-116c-4ffb-a70b-e5e35d01e3b1">

GenevieveBuckley commented:
> About the strategy here: what I would suggest is to first clean up the changes to the installation, fix conflicts and provide the sample data plugins in this PR. And then work on the 2d plugin in a follow up PR.

Agreed, this sounds sensible.

I have:
- [x] Cleaned up the changes to the installation (we still have the black and ruff config settings in setup.cfg - do you want them to stay or be removed?)
- [x] Fixed the merge conflicts
- [x] Provided all the sample datasets as part of the plugin (including image-series which now opens three separate layers - it just felt too weird to leave one dataset out, and also the new-ish segmentation dataset hosted at OwnCloud) 
- [x] Checked the existing test suite passes both locally and on CI

constantinpape commented:
> You access the sample datasets from the napari File menu (but plugin widgets are accessed from the plugins menu).
> 
> File > Open Sample > SegmentAnything for Microscopy > ...

Thanks, that works!

constantinpape reviewed (approved):
Thanks @GenevieveBuckley! I checked it out and it's working as expected. We can leave `black` and `ruff` in; at some point I want to have CI to check lint in the commits, and this could be a good starting point then.
I will follow up on details for the implementation of the 2d annotator later.
